packages/contracts-bedrock/src/L1/OPContractsManager.sol

```diff
@@ -363,6 +374,40 @@ contract OPContractsManager is ISemver {
         return output;
     }
 
+    /// @notice Upgrades a set of chains to the latest implementation contracts
+    /// @param _systemConfigs Array of SystemConfig contracts, one per chain to upgrade
+    /// @param _proxyAdmins Array of ProxyAdmin contracts, one per chain to upgrade
+    /// @dev This function is intended to be called via DELEGATECALL from the Upgrade Controller Safe
+    function upgrade(ISystemConfig[] calldata _systemConfigs, IProxyAdmin[] calldata _proxyAdmins) external {
+        // TODO: emit the Upgraded event
+        // TODO: enforce delegatecalling
+        Implementations memory impls = thisOPCM.implementations();
+
+        for (uint256 i = 0; i < _systemConfigs.length; i++) {
+            ISystemConfig systemConfig = _systemConfigs[i];
+            ISystemConfig.Addresses memory opChainAddrs = ISystemConfig.Addresses({
+                l1CrossDomainMessenger: systemConfig.l1CrossDomainMessenger(),
+                l1ERC721Bridge: systemConfig.l1ERC721Bridge(),
+                l1StandardBridge: systemConfig.l1StandardBridge(),
+                disputeGameFactory: systemConfig.disputeGameFactory(),
+                optimismPortal: systemConfig.optimismPortal(),
+                optimismMintableERC20Factory: systemConfig.optimismMintableERC20Factory(),
+                gasPayingToken: address(0)
+            });
+
+            IProxyAdmin proxyAdmin = _proxyAdmins[i];
+            proxyAdmin.upgrade(payable(address(systemConfig)), impls.systemConfigImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.l1CrossDomainMessenger), impls.l1CrossDomainMessengerImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.l1ERC721Bridge), impls.l1ERC721BridgeImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.l1StandardBridge), impls.l1StandardBridgeImpl);
+            proxyAdmin.upgrade(payable(opChainAddrs.disputeGameFactory), impls.disputeGameFactoryImpl);
```

Comment:
Direction:
1. we _think_ that this should also discover and update the implementations for the ASR and DelayedWeth (but not the dispute games themselves).
2. Implement that in a follow up PR, and have the proofs team review.

Reply:
added a todo